diff --git a/types/wlr_output_layout.c b/types/wlr_output_layout.c
index 28d91e8d42cfb75b9ca151be451e5ba287f33ca0..ef6e2850ca0e81d12b1cac9d1bc059b0f1f93694 100644
--- a/types/wlr_output_layout.c
+++ b/types/wlr_output_layout.c
@@ -231,6 +231,9 @@ bool wlr_output_layout_contains_point(struct wlr_output_layout *layout,
 	if (reference) {
 		struct wlr_output_layout_output *l_output =
 			wlr_output_layout_get(layout, reference);
+		if (!l_output) {
+			return false;
+		}
 		struct wlr_box output_box;
 		output_layout_output_get_box(l_output, &output_box);
 		return wlr_box_contains_point(&output_box, lx, ly);
