From c128b1be5f571315ae8a0f34cd9d00fd20c30cdc Mon Sep 17 00:00:00 2001
From: Adrien Plazas <kekun.plazas@laposte.net>
Date: Mon, 27 Apr 2020 10:42:16 +0200
Subject: [PATCH 2/8] accounts-editor-add-pane: Drop the useless shadow

---
 ui/accounts_editor_add_pane.ui | 1 -
 1 file changed, 1 deletion(-)

diff --git a/ui/accounts_editor_add_pane.ui b/ui/accounts_editor_add_pane.ui
index 0c01a4c1..336b73ea 100644
--- a/ui/accounts_editor_add_pane.ui
+++ b/ui/accounts_editor_add_pane.ui
@@ -89,7 +89,6 @@
         <property name="vexpand">True</property>
         <property name="vadjustment">pane_adjustment</property>
         <property name="hscrollbar_policy">never</property>
-        <property name="shadow_type">in</property>
         <child>
           <object class="GtkViewport">
             <property name="visible">True</property>
-- 
2.29.2

