From ee3e56aa1f3a71457b43a79b233ba3200e4e78d2 Mon Sep 17 00:00:00 2001
From: Julian Sparber <julian@sparber.net>
Date: Mon, 12 Oct 2020 16:35:22 +0200
Subject: [PATCH 3/6] conversation-viewer: allow one email per line

---
 ui/conversation-message.ui | 6 ------
 1 file changed, 6 deletions(-)

diff --git a/ui/conversation-message.ui b/ui/conversation-message.ui
index 95560939..d68817b2 100644
--- a/ui/conversation-message.ui
+++ b/ui/conversation-message.ui
@@ -157,7 +157,6 @@
                                 <property name="valign">baseline</property>
                                 <property name="hexpand">True</property>
                                 <property name="column_spacing">2</property>
-                                <property name="min_children_per_line">1</property>
                                 <property name="max_children_per_line">4</property>
                                 <property name="selection_mode">none</property>
                                 <signal name="child-activated" handler="on_address_box_child_activated" swapped="no"/>
@@ -218,7 +217,6 @@
                                 <property name="valign">start</property>
                                 <property name="hexpand">True</property>
                                 <property name="column_spacing">2</property>
-                                <property name="min_children_per_line">2</property>
                                 <property name="max_children_per_line">4</property>
                                 <property name="selection_mode">none</property>
                                 <signal name="child-activated" handler="on_address_box_child_activated" swapped="no"/>
@@ -263,7 +261,6 @@
                                 <property name="valign">start</property>
                                 <property name="hexpand">True</property>
                                 <property name="column_spacing">2</property>
-                                <property name="min_children_per_line">2</property>
                                 <property name="max_children_per_line">4</property>
                                 <property name="selection_mode">none</property>
                                 <signal name="child-activated" handler="on_address_box_child_activated" swapped="no"/>
@@ -328,7 +325,6 @@
                                 <property name="valign">start</property>
                                 <property name="hexpand">True</property>
                                 <property name="column_spacing">2</property>
-                                <property name="min_children_per_line">2</property>
                                 <property name="max_children_per_line">4</property>
                                 <property name="selection_mode">none</property>
                                 <signal name="child-activated" handler="on_address_box_child_activated" swapped="no"/>
@@ -373,7 +369,6 @@
                                 <property name="valign">start</property>
                                 <property name="hexpand">True</property>
                                 <property name="column_spacing">2</property>
-                                <property name="min_children_per_line">2</property>
                                 <property name="max_children_per_line">4</property>
                                 <property name="selection_mode">none</property>
                                 <signal name="child-activated" handler="on_address_box_child_activated" swapped="no"/>
@@ -418,7 +413,6 @@
                                 <property name="valign">start</property>
                                 <property name="hexpand">True</property>
                                 <property name="column_spacing">2</property>
-                                <property name="min_children_per_line">2</property>
                                 <property name="max_children_per_line">4</property>
                                 <property name="selection_mode">none</property>
                                 <signal name="child-activated" handler="on_address_box_child_activated" swapped="no"/>
-- 
2.29.2

