From d97d86cbc73dc0dc565494c1083b5052457afe04 Mon Sep 17 00:00:00 2001
From: Adrien Plazas <kekun.plazas@laposte.net>
Date: Wed, 16 Sep 2020 13:19:00 +0200
Subject: [PATCH 057/124] toolbar header group

---
 ui/main-toolbar.ui | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/ui/main-toolbar.ui b/ui/main-toolbar.ui
index 871aba25..5d58af6f 100644
--- a/ui/main-toolbar.ui
+++ b/ui/main-toolbar.ui
@@ -425,10 +425,17 @@
       </packing>
     </child>
   </template>
-  <object class="HdyHeaderGroup" id="header_group">
+  <object class="HdyHeaderGroup" id="conversations_header_group">
+    <property name="decorate-all" bind-source="conversations_leaflet" bind-property="folded" bind-flags="sync-create"/>
     <headerbars>
       <headerbar name="folder_header"/>
       <headerbar name="conversations_header"/>
+    </headerbars>
+  </object>
+  <object class="HdyHeaderGroup" id="header_group">
+    <property name="decorate-all" bind-source="MainToolbar" bind-property="folded" bind-flags="sync-create"/>
+    <headerbars>
+      <headerbar name="conversations_header_group"/>
       <headerbar name="conversation_header"/>
     </headerbars>
   </object>
-- 
2.29.2

