From 4be51e01c02f57fe68121603888ca012000d845d Mon Sep 17 00:00:00 2001
From: Michael Gratton <mike@vee.net>
Date: Sat, 10 Oct 2020 23:57:11 +1100
Subject: [PATCH 064/124] Util.Date: Remove now unused function

---
 src/client/util/util-date.vala | 28 ----------------------------
 1 file changed, 28 deletions(-)

diff --git a/src/client/util/util-date.vala b/src/client/util/util-date.vala
index 88948916..322a4cf2 100644
--- a/src/client/util/util-date.vala
+++ b/src/client/util/util-date.vala
@@ -208,32 +208,4 @@ public string get_clock_format(ClockFormat clock_format) {
     return xlat_pretty_clocks[clock_format.to_index()];
 }
 
-public string get_full_date(ClockFormat clock_format) {
-    var value = "";
-    switch (clock_format) {
-    case TWELVE_HOURS:
-        /// 12 hours format for datetime that a message being replied
-        /// to was received See
-        /// http://developer.gnome.org/glib/2.32/glib-GDateTime.html#g-date-time-format
-        value = _("%a, %b %-e, %Y at %l:%M %P");
-        break;
-    case TWENTY_FOUR_HOURS:
-        /// 24 hours format for the datetime that a message being
-        /// replied to was received See
-        /// http://developer.gnome.org/glib/2.32/glib-GDateTime.html#g-date-time-format
-        value = _("%a, %b %-e, %Y at %H:%M");
-        break;
-    case LOCALE_DEFAULT:
-        /// Format for the datetime that a message being replied to
-        /// was received See
-        /// http://developer.gnome.org/glib/2.32/glib-GDateTime.html#g-date-time-format
-        value = _("%a, %b %-e, %Y at %X");
-        break;
-    case TOTAL:
-        // noop
-        break;
-    }
-    return value;
-}
-
 }
-- 
2.29.2

