From 3f7c054db085a846b550c2cbc9fb71b352add2b9 Mon Sep 17 00:00:00 2001
From: Michael Gratton <mike@vee.net>
Date: Sat, 29 Aug 2020 12:57:08 +1000
Subject: [PATCH 084/124] build: Bump WebKitGTK min version to include
 UserMessage support

---
 meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 240eacaa..9b55b24d 100644
--- a/meson.build
+++ b/meson.build
@@ -53,7 +53,7 @@ valac = meson.get_compiler('vala')
 target_vala = '0.48.6'
 target_glib = '2.64'
 target_gtk = '3.24.7'
-target_webkit = '2.26'
+target_webkit = '2.28'
 
 if not valac.version().version_compare('>=' + target_vala)
   error('Vala does not meet minimum required version: ' + target_vala)
-- 
2.29.2

