From: Michael Gratton <mike@vee.net>
Date: Fri, 11 Oct 2019 13:42:17 +1100
Subject: Bump client-test timeout to 300s

---
 test/meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

Index: geary-3.38.0.1/test/meson.build
===================================================================
--- geary-3.38.0.1.orig/test/meson.build
+++ geary-3.38.0.1/test/meson.build
@@ -168,4 +168,4 @@ test_integration_bin = executable('test-
 )
 
 test('engine-tests', test_engine_bin)
-test('client-tests', test_client_bin)
+test('client-tests', test_client_bin, timeout: 300)
