diff --git a/daemon/gdm-session-worker.c b/daemon/gdm-session-worker.c
index 88882380b3be47580aad70220fd3c7e1da1b08e1..b062fdd7f4e597d2784a99290a2704dcaf81f054 100644
--- a/daemon/gdm-session-worker.c
+++ b/daemon/gdm-session-worker.c
@@ -3572,7 +3572,7 @@ gdm_session_worker_finalize (GObject *object)
 
         jump_back_to_initial_vt (worker);
 
-        g_object_unref (worker->user_settings);
+        g_clear_object (&worker->user_settings);
         g_free (worker->service);
         g_free (worker->x11_display_name);
         g_free (worker->x11_authority_file);
