--- a/src/rotation-manager.c	2020-08-06 13:39:48.934458000 +0200
+++ b/src/rotation-manager.c	2020-08-06 13:40:52.711124361 +0200
@@ -79,11 +79,11 @@ match_orientation (PhoshRotationManager
 
   if (!g_strcmp0 ("normal", orient)) {
     transform = PHOSH_MONITOR_TRANSFORM_NORMAL;
-  } else if (!g_strcmp0 ("right-up", orient)) {
+  } else if (!g_strcmp0 ("left-up", orient)) {
     transform = PHOSH_MONITOR_TRANSFORM_270;
   } else if (!g_strcmp0 ("bottom-up", orient)) {
     transform = PHOSH_MONITOR_TRANSFORM_180;
-  } else if (!g_strcmp0 ("left-up", orient)) {
+  } else if (!g_strcmp0 ("right-up", orient)) {
     transform = PHOSH_MONITOR_TRANSFORM_90;
   } else if (!g_strcmp0 ("undefined", orient)) {
     return; /* just leave as is */
