From 41867cdcb20661ae11ac7f85eb367685978fcf59 Mon Sep 17 00:00:00 2001
From: Mohammed Sadiq <sadiq@sadiqpk.org>
Date: Wed, 28 Apr 2021 12:30:49 +0530
Subject: [PATCH] top-panel: Update power button and popover styles to match
 design

---
 src/ui/top-panel.ui | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/src/ui/top-panel.ui b/src/ui/top-panel.ui
index 5dd7ca7a..9135b053 100644
--- a/src/ui/top-panel.ui
+++ b/src/ui/top-panel.ui
@@ -3,6 +3,9 @@
 <interface>
   <requires lib="gtk+" version="3.20"/>
   <object class="GtkPopoverMenu" id="menu_power">
+    <style>
+      <class name="osd"/>
+    </style>
     <child>
       <object class="GtkBox">
         <property name="visible">True</property>
@@ -260,8 +263,11 @@
                     <property name="can_focus">True</property>
                     <property name="receives_default">True</property>
                     <property name="popover">menu_power</property>
+                    <property name="width-request">32</property>
+                    <property name="height-request">32</property>
                     <style>
                       <class name="phosh-power-button"/>
+                      <class name="circular"/>
                     </style>
                     <child>
                       <object class="GtkImage">
-- 
GitLab

