diff --git a/src/polkit-auth-agent.h b/src/polkit-auth-agent.h
index fe7e808..88b8a19 100644
--- a/src/polkit-auth-agent.h
+++ b/src/polkit-auth-agent.h
@@ -19,7 +19,11 @@ G_BEGIN_DECLS
 #if POLKIT_AGENT_MAJOR_VERSION == 0 && POLKIT_AGENT_MINOR_VERSION < 114
 G_DEFINE_AUTOPTR_CLEANUP_FUNC(PolkitSubject, g_object_unref)
 #endif
+
+/* libpolkitAgentListener lacks these */
+#if POLKIT_AGENT_MAJOR_VERSION == 0 && POLKIT_AGENT_MINOR_VERSION < 120
 G_DEFINE_AUTOPTR_CLEANUP_FUNC(PolkitAgentListener, g_object_unref)
+#endif
 
 #define PHOSH_TYPE_POLKIT_AUTH_AGENT (phosh_polkit_auth_agent_get_type())
