--- linux-pine64-5.9-da1a1dee4471aa5da391517c4af40d2c80fe828d/drivers/media/i2c/ov5640.c	2020-10-18 00:13:18.000000000 +0200
+++ linux-pine64-5.9-da1a1dee4471aa5da391517c4af40d2c80fe828d/drivers/media/i2c/ov5640.c	2020-10-18 22:19:32.490533074 +0200
@@ -2677,7 +2677,7 @@ static int ov5640_set_ctrl_focus(struct
 	if (ret) {
 		dev_err(&client->dev, "%s: no autofocus firmware loaded\n",
 			__func__);
-		return ret;
+		return 0;
 	}
 
 	if (command == OV5640_FW_CMD_RELEASE_FOCUS) {
