--- azure-cli-2.21.0.a/azure/cli/command_modules/util/commands.py	2021-03-19 06:54:29.000000000 +0100
+++ azure-cli-2.21.0.b/azure/cli/command_modules/util/commands.py	2021-04-05 16:50:42.425805243 +0200
@@ -12,8 +12,5 @@
     with self.command_group('') as g:
         g.custom_command('version', 'show_version')
 
-    with self.command_group('') as g:
-        g.custom_command('upgrade', 'upgrade_version', is_preview=True)
-
     with self.command_group('demo', deprecate_info=g.deprecate(hide=True)) as g:
         g.custom_command('style', 'demo_style')
