From 4a657f542df799434eba9163861109b8a4996bb9 Mon Sep 17 00:00:00 2001
From: Gerben Jan Dijkman <gjdijkman@gjdwebserver.nl>
Date: Mon, 11 Jul 2022 12:09:07 +0200
Subject: [PATCH] Removing online deps since they keep breaking

---
 Cargo.deps.online | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Cargo.deps.online b/Cargo.deps.online
index 54e8e17..c7d3adb 100644
--- a/Cargo.deps.online
+++ b/Cargo.deps.online
@@ -1,4 +1,4 @@
 # Dependencies which are only used with online, crates.io builds.
 [patch.crates-io]
 # Dependency was yanked, but gio 0.7 needs it.
-fragile = { git = "https://source.puri.sm/dorota.czaplejewicz/fragile.git", tag = "0.3.0" }
+# fragile = { git = "https://source.puri.sm/dorota.czaplejewicz/fragile.git", tag = "0.3.0" }
-- 
2.35.1

